### PR TITLE
Simplify RewardsEngine::GetExternalWallet

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/34063

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_rewards/core/wallet_provider/connect_external_wallet.cc [hunk 1]
```diff
@@ -177,18 +177,27 @@ void ConnectExternalWallet::OnConnect(
...
-      WalletType() + std::string("/") + abbreviated_address);
+  // Set the "active" wallet type.
+  engine_->SetState(state::kExternalWalletType, std::string(WalletType()));
 
   // Update the user's "declared country" based on the information provided by
   // the linking endpoint.
   CHECK(result.has_value() && !result.value().empty());
   engine_->SetState(state::kDeclaredGeo, result.value());
 
+  // Attempt to transfer any upspent vBAT to the user's external wallet.
```
Comment (emerick) [inline components/brave_rewards/core/wallet_provider/connect_external_wallet.cc:188]: Typo: unspent

**Review by emerick — APPROVED** [review]

File: components/brave_rewards/core/rewards_engine_impl.cc [hunk 2]
```diff
@@ -577,27 +578,33 @@ void RewardsEngineImpl::FetchBalance(FetchBalanceCallback callback) {
...
 }
 
-void RewardsEngineImpl::GetExternalWallet(const std::string& wallet_type,
-                                          GetExternalWalletCallback callback) {
-  WhenReady([this, wallet_type, callback = std::move(callback)]() mutable {
-    if (wallet_type == constant::kWalletBitflyer) {
-      return bitflyer()->GetWallet(std::move(callback));
+void RewardsEngineImpl::GetExternalWallet(GetExternalWalletCallback callback) {
+  WhenReady([this, callback = std::move(callback)]() mutable {
+    auto wallet_type = GetState<std::string>(state::kExternalWalletType);
```
Comment (szilardszaloki) [inline components/brave_rewards/core/rewards_engine_impl.cc:583]: For profiles created before July 2021 (this is when Gemini was added), the `external_wallet_type` pref doesn't exist for bitFlyer and Uphold users. For them the `GetExternalWallet()` flow would return `nullptr`.

Comment (zenparsing) [reply components/brave_rewards/core/rewards_engine_impl.cc:583]: Thanks for the reminder! I'll add the necessary logic (possibly a migration) and document.

File: components/brave_rewards/core/wallet_provider/connect_external_wallet.cc [hunk 3]
```diff
@@ -177,18 +177,27 @@ void ConnectExternalWallet::OnConnect(
...
+  // Set the "active" wallet type.
+  engine_->SetState(state::kExternalWalletType, std::string(WalletType()));
 
   // Update the user's "declared country" based on the information provided by
   // the linking endpoint.
   CHECK(result.has_value() && !result.value().empty());
   engine_->SetState(state::kDeclaredGeo, result.value());
 
+  // Attempt to transfer any unspent vBAT to the user's external wallet.
+  engine_->promotion()->TransferTokens(base::DoNothing());
```
Comment (zenparsing) [inline components/brave_rewards/core/wallet_provider/connect_external_wallet.cc:189]: Moved from the "get wallet" flow. @szilardszaloki Does it seem fine to move it here (to the connect wallet flow), or was there a particular reason for putting it in the "get wallet" flow?

Comment (szilardszaloki) [reply components/brave_rewards/core/wallet_provider/connect_external_wallet.cc:189]: @zenparsing The reason it was in the "get wallet" flow is that this flow runs more frequently than any other flows.

File: components/brave_rewards/core/rewards_engine_impl.cc (hunk 2 above)
Comment (zenparsing) [reply components/brave_rewards/core/rewards_engine_impl.cc:583]: Migration added

File: components/brave_rewards/core/wallet_provider/connect_external_wallet.cc (hunk 3 above)
Comment (zenparsing) [reply components/brave_rewards/core/wallet_provider/connect_external_wallet.cc]: The transfer endpoint is no longer supported, so we'll remove the call.

File: components/brave_rewards/core/state/state_migration_v14.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+// Migration 14 corrects the situation where the `external_wallet_type` pref is
+// empty but the user has a connected wallet. Users that connected before
+// `external_wallet_type` was introduced may be in this state.
+class StateMigrationV14 {
+ public:
+  explicit StateMigrationV14(RewardsEngineImpl& engine);
+  ~StateMigrationV14();
+
+  void Migrate(LegacyResultCallback);
```
Comment (szilardszaloki) [inline components/brave_rewards/core/state/state_migration_v14.h:27]: Would you mind using `ResultCallback` instead?

File: components/brave_rewards/core/wallet_provider/wallet_provider.cc [hunk 5]
```diff
@@ -120,15 +120,14 @@ void WalletProvider::BeginLogin(BeginExternalWalletLoginCallback callback) {
 void WalletProvider::ConnectWallet(
     const base::flat_map<std::string, std::string>& args,
     ConnectExternalWalletCallback callback) {
+  // The connect wallet flow currently requires an ExternalWallet structure.
+  // Create one now if necessary.
+  wallet::MaybeCreateWallet(*engine_, WalletType());
```
Comment (szilardszaloki) [inline components/brave_rewards/core/wallet_provider/wallet_provider.cc:125]: If this call is necessary here, then I think we're doing something wrong. Is there a specific use case where this is needed? Also, the connect wallet flow should be able to handle the cases where there's no external wallet created yet - it should error out with an "unexpected" error code.

Comment (zenparsing) [reply components/brave_rewards/core/wallet_provider/wallet_provider.cc:125]: I was surprised by this as well, but it makes sense given these refactorings. Previously, we needed a `kNotConnected` wallet in order to transmit the login URL to the front-end. By the time we got into this code, the wallet was sure to be created. Without this call to create a wallet, the connect flow will error out as soon as it tries to get the wallet data. For instance:

```cpp
  if (!engine_->uphold()->GetWalletIf({mojom::WalletStatus::kNotConnected,
                                       mojom::WalletStatus::kLoggedOut})) {
    return std::move(callback).Run(ConnectExternalWalletResult::kUnexpected);
  }
```

**Review by szilardszaloki — APPROVED** [review]

**Pull request merged**